org.eclipse.dartboard/src/org/eclipse/dartboard/launch/LaunchShortcut.java

```diff
@@ -38,57 +45,77 @@
  * @author Jonas Hungershausen
  *
  */
+@SuppressWarnings("restriction")
 public class LaunchShortcut implements ILaunchShortcut {
 
 	@Override
 	public void launch(ISelection selection, String mode) {
+		// Get the active project from the
 		IProject selected = null;
 		if (selection instanceof StructuredSelection) {
 			Object firstElement = ((StructuredSelection) selection).getFirstElement();
-			if (firstElement instanceof IProject) {
-				selected = (IProject) firstElement;
+			if (firstElement instanceof IAdaptable) {
+				selected = ((IAdaptable) firstElement).getAdapter(IProject.class);
 			}
 		}
 
+		launchProject(selected, mode);
+	}
+
+	@Override
+	public void launch(IEditorPart editor, String mode) {
+		IResource resource = editor.getEditorInput().getAdapter(IResource.class);
+
+		if (resource == null) {
+			MessageDialog.openError(null, Messages.Launch_ConfigurationRequired_Title,
+					Messages.Launch_ConfigurationRequired_Body);
+			return;
+		}
+
+		launchProject(resource.getProject(), mode);
+	}
+
+	private void launchProject(IProject project, String mode) {
+		if (project == null) {
+			MessageDialog.openError(null, Messages.Launch_ConfigurationRequired_Title,
+					Messages.Launch_ConfigurationRequired_Body);
+			return;
+		}
+
 		ILaunchManager manager = DebugPlugin.getDefault().getLaunchManager();
 
 		ILaunchConfigurationType type = manager.getLaunchConfigurationType(Constants.LAUNCH_CONFIGURATION_ID);
 
 		Set<String> modes = new HashSet<>();
```

Comment:
Are we using the Set ?. If not remove it

Reply:
Oh, good catch. Relic from before I think.